// vite-plugin-cdn/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
const createCdnReplacePlugin_1 = require("./createCdnReplacePlugin");
exports.default = (opt) => {
    const { enabled = process.env.NODE_ENV === 'production', modules = [], cssInject = 'body', esm = true, } = opt || {};
    if (!enabled) {
        return {};
    }
    const external = modules.map((item) => item.name);
    const cssList = modules.map((item) => item.css).filter(Boolean);
    const jsList = modules.map((item) => item.url).filter(Boolean);
    return {
        indexHtmlTransforms: [
            {
                apply: 'post',
                transform: ({ code, isBuild }) => {
                    if (!isBuild)
                        return code;
                    let processCode = utils_1.injectCss(code, cssList, cssInject);
                    if (!esm) {
                        processCode = utils_1.injectJs(code, jsList);
                    }
                    return processCode;
                },
            },
        ],
        rollupInputOptions: {
            external,
        },
        rollupOutputOptions: {
            plugins: [createCdnReplacePlugin_1.createCdnReplacePlugin(opt)],
        },
    };
};
